src/esssans/loki/masking.py

```diff
@@ -0,0 +1,117 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+"""
+Masking functions for the loki workflow.
+"""
+from typing import NewType, Optional
+
+import scipp as sc
+
+from ..types import (
+    CalibratedMaskedData,
+    CleanMasked,
+    DataWithLogicalDims,
+    MaskedData,
+    Numerator,
+    RunType,
+    SampleRun,
+)
+
+DetectorLowCountsStrawMask = NewType('DetectorLowCountsStrawMask', sc.Variable)
+"""Detector low-counts straw mask"""
+DetectorBadStrawsMask = NewType('DetectorBadStrawsMask', sc.Variable)
+"""Detector bad straws mask"""
+DetectorBeamStopMask = NewType('DetectorBeamStopMask', sc.Variable)
+"""Detector beam stop mask"""
+DetectorTubeEdgeMask = NewType('DetectorTubeEdgeMask', sc.Variable)
+"""Detector tube edge mask"""
+
+
+def detector_straw_mask(
+    sample_straws: CalibratedMaskedData[SampleRun],
+) -> DetectorLowCountsStrawMask:
+    return DetectorLowCountsStrawMask(
+        sample_straws.sum(['tof', 'pixel']).data < sc.scalar(300.0, unit='counts')
+    )
+
+
+def detector_beam_stop_mask(
+    sample_straws: CalibratedMaskedData[SampleRun],
+) -> DetectorBeamStopMask:
+    pos = sample_straws.coords['position'].copy()
+    pos.fields.z *= 0.0
+    return DetectorBeamStopMask((sc.norm(pos) < sc.scalar(0.042, unit='m')))
+
+
+def detector_tube_edge_mask(
+    sample_straws: CalibratedMaskedData[SampleRun],
+) -> DetectorTubeEdgeMask:
+    return DetectorTubeEdgeMask(
+        (abs(sample_straws.coords['position'].fields.x) > sc.scalar(0.36, unit='m'))
+        | (abs(sample_straws.coords['position'].fields.y) > sc.scalar(0.28, unit='m'))
+    )
+
+
+def mask_detectors(
+    da: DataWithLogicalDims[RunType],
+) -> MaskedData[RunType]:
+    """Apply pixel-specific masks to raw data.
+
+    Parameters
+    ----------
+    da:
+        Raw data.
+    """
+    # Beam stop
+    da = da.copy(deep=False)
+    counts = da.sum('tof').data
+    r = sc.sqrt(
+        da.coords['position'].fields.x ** 2 + da.coords['position'].fields.y ** 2
+    )
+    da.masks['low_counts_middle'] = (counts < sc.scalar(20.0, unit='counts')) & (
+        r < sc.scalar(0.075, unit='m')
+    )
+    # Low counts
+    da.masks['very_low_counts'] = counts < sc.scalar(3.0, unit='counts')
```

Comment:
Are you sure about this? I don't think we can just mask pixels with low counts, since they are very important for larger Q?

Reply:
This relies on something that you pointed out further down which relies on subsequently removing the masks that were used to find the beam center and then adding some new masks back in.

I am starting to doubt the strategy of having some masking before beam center finding, and some other masking after.
I did this because masking the edges of the panel was much easier programmatically after beam center was applied (just an `abs(x)` or `abs(y)`.
But I think I want to go back to the way it was in `main`, with only masking once before beam center finding?